### PR TITLE
[FEATURE] Introduce PSR-14 ModifyRecordOverlayIconIdentifierEvent #629

### PR DESCRIPTION
**Pull request by FlorianT03**
resolves: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/629

releases: main

### CODE REVIEW

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+extension authors to modify the overlay icon identifier of any record icon.
+
+Extensions can listen to the :php:`ModifyRecordOverlayIconIdentifierEvent`
+and perform necessary modifications to the overlay icon identifier
+based on their requirements.
+
+Example
+=======
+
+The corresponding event listener class:
```
Comment (linawolf) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:27]: The event listener class is corresponding to what?

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+:php:`\TYPO3\CMS\Core\Imaging\Event\ModifyRecordOverlayIconIdentifierEvent` allows
+extension authors to modify the overlay icon identifier of any record icon.
+
+Extensions can listen to the :php:`ModifyRecordOverlayIconIdentifierEvent`
+and perform necessary modifications to the overlay icon identifier
+based on their requirements.
+
+Example
+=======
+
```
Comment (linawolf) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:26]: Please also add an example for the Services.yaml (see other events with examples)

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+The corresponding event listener class:
+
+..  literalinclude:: _ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php
+    :language: php
+    :caption: EXT:my_extension/Classes/Core/EventListener/ModifyRecordOverlayIconIdentifierEventListener.php
+
+API
+===
+
+..  include:: /CodeSnippets/Events/Core/ModifyRecordOverlayIconIdentifierEvent.rst.txt
```
Comment (linawolf) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:36]: There must always be an empty line at the end of each file. You can configure your PHP storm to do that automtically

**Comment by brotkrueml** [comment]
@FlorianT03 This PR mixes several changes that are not connected to the new feature. Please hold the changeset small and provide only the new event page (and necessary changes for that, like the index page).

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 4]
```diff
@@ -0,0 +1,36 @@
+..  include:: /Includes.rst.txt
+..  index:: Events; ModifyRecordOverlayIconIdentifierEvent
+..  _ModifyRecordOverlayIconIdentifierEvent:
+
+======================================
+ModifyRecordOverlayIconIdentifierEvent
+======================================
+..  versionadded:: 13.0
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:8]: Please add an empty line before the directive.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    The PSR-14 event :php:`ModifyRecordOverlayIconIdentifierEvent` has
+    been introduced which serves as a more flexible replacement for the
+    removed hook:
+
+    *   :php:`$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['TYPO3\CMS\Core\Imaging\IconFactory']['overrideIconOverlay']`
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:14]: The list is not necessary as it is only one item, so I suggest to write:

    removed hook :php:`...`

without the list.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+The PSR-14 event
+:php:`\TYPO3\CMS\Core\Imaging\Event\ModifyRecordOverlayIconIdentifierEvent` allows
+extension authors to modify the overlay icon identifier of any record icon.
+
+Extensions can listen to the :php:`ModifyRecordOverlayIconIdentifierEvent`
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:20]: This paragraph is mostly redundant to the one above. Please combine them.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst (hunk 2 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:26]: We don't need the Services.yaml to tag event listeners anymore when using the newly introduced PHP attributes, see also #3382.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst (hunk 1 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:27]: This sentence should be omitted like for other events in v13, for example:
https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Events/Events/Backend/AfterBackendPageRenderEvent.html

**Review by brotkrueml — CHANGES_REQUESTED** [review]
I suggest to add a hint that the event listener can also be tagged via the Services.yaml, like here:
https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Events/Events/Backend/AfterBackendPageRenderEvent.html#example

However, as this is a new event, I would rephrase it a little bit for new events:

    ..  versionadded:: 13.0
        The PHP attribute :php:`\TYPO3\CMS\Core\Attribute\AsEventListener` has been
        introduced to tag a PHP class as an event listener. Alternatively, you can also
        register an event listener via the :file:`Configuration/Services.yaml` file. Have
        a look into the section :ref:`EventDispatcherImplementation`.

So, you can provide a new include file for those new events.

What do you think?

Edit: I added a suggestion to include the now available include.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/Index.rst [hunk 7]
```diff
@@ -0,0 +1,19 @@
+..  include:: /Includes.rst.txt
+..  index:: pair: Events; Core Imaging
+..  _eventlist-core-page:
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/Index.rst:3]:
```suggestion
..  _eventlist-core-imaging:
```

File: Documentation/ApiOverview/Events/Events/Core/Imaging/_ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Vendor\MyPackage\Core\EventListener;
+
+use TYPO3\CMS\Core\Attribute\AsEventListener;
+use TYPO3\CMS\Core\Imaging\Event\ModifyRecordOverlayIconIdentifierEvent;
+
+final class ModifyRecordOverlayIconIdentifierEventListener
+{
+    #[AsEventListener('my-package/core/modify-record-overlay-icon-identifier')]
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/_ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php:10]:
```suggestion
    #[AsEventListener('my-extension/imaging/modify-record-overlay-icon-identifier')]
```

File: Documentation/ApiOverview/Events/Events/Core/Imaging/_ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Vendor\MyPackage\Core\EventListener;
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/_ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php:3]:
```suggestion
namespace Vendor\MyPackage\Imaging\EventListener;
```

This is a better namespace than "Core"

File: Documentation/ApiOverview/Events/Events/Core/Imaging/_ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/_ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php:1]:
```suggestion
<?php

declare(strict_types=1);
```

**Comment by linawolf** [comment]
> @FlorianT03 This PR mixes several changes that are not connected to the new feature. Please hold the changeset small and provide only the new event page (and necessary changes for that, like the index page).

I removed those, my fault.

File: Documentation/ApiOverview/RequestLifeCycle/RequestAttributes/CurrentContentObject.rst [hunk 11]
```diff
@@ -0,0 +1,32 @@
+..  include:: /Includes.rst.txt
```
Comment (brotkrueml) [inline Documentation/ApiOverview/RequestLifeCycle/RequestAttributes/CurrentContentObject.rst:1]: This file is not related to the event.

File: Documentation/CodeSnippets/Events/Core/ModifyRecordOverlayIconIdentifierEvent.rst.txt [hunk 12]
```diff
@@ -0,0 +1,27 @@
+.. Generated by https://github.com/TYPO3-Documentation/t3docs-codesnippets
```
Comment (brotkrueml) [inline Documentation/CodeSnippets/Events/Core/ModifyRecordOverlayIconIdentifierEvent.rst.txt:1]: Please regenerate the snippets after https://github.com/TYPO3-Documentation/codesnippet/pull/20 was merged and update your snippets project accordingly.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+..  index:: Events; ModifyRecordOverlayIconIdentifierEvent
+..  _ModifyRecordOverlayIconIdentifierEvent:
+
+======================================
+ModifyRecordOverlayIconIdentifierEvent
+======================================
+
+..  versionadded:: 13.0
+
+The PSR-14 event `ModifyRecordOverlayIconIdentifierEvent` has been introduced,
```
[the remark this answers is not in the document]
Comment (linawolf) [reply Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:11]: @FlorianT03 you can use the local history in PHP storm like I showed you to get the list with the hook back and then transform it into a sentence.

**Comment by brotkrueml** [comment]
Hej @FlorianT03, can be get this PR to a state so it can be merged? What is missing to make the adjustments? Do you need help?

**Comment by FlorianT03** [comment]
@brotkrueml Well, I thought I finished this PR (which is not the case) and then forgot about it tbh. It's because I was writing some documentation for the sys_note extension and was correcting some other stuff in the meantime.. Maybe I should not begin 5 things at once.. But this PR is now on top of my To-Do List :D

**Comment by brotkrueml** [comment]
Set on hold, will make a PR of another new PSR-14 event with the proposed include. Then this one can be tackled again.

File: Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+to the overlay icon identifier based on their requirements.
+
+
+Example
+=======
+
+..  literalinclude:: _ModifyRecordOverlayIconIdentifierEvent/_ModifyRecordOverlayIconIdentifierEventListener.php
+    :language: php
+    :caption: EXT:my_extension/Classes/Core/EventListener/ModifyRecordOverlayIconIdentifierEventListener.php
+
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Events/Events/Core/Imaging/ModifyRecordOverlayIconIdentifierEvent.rst:26]:
```suggestion

..  include:: /_includes/EventsAttributeAddedNew.rst.txt
```

Please merge main into your branch to get the include introduced with #3520.

**Review by linawolf — APPROVED** [review]

**Review by brotkrueml — APPROVED** [review]
